### PR TITLE
Code Coverage Updates

### PR DESCRIPTION
**Pull request by reidkwja**
The following PR will automate the coverage report in addition to fixing prior issues

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -117,12 +117,21 @@ In order for a PR to be accepted the following pylint command needs to result in
...
 ```
-## Coverage 
+## Code Coverage 
 
-To extract coverage percentage from the underlying code base prior to opening Pull Request:
+To manually extract coverage percentage using the command line:
 
 ```
-cd MITunaX
+//run the docker image and mount the MITuna repo folder to it from host
```
Comment (alexandraBara) [inline README.md]:
```suggestion
//run the docker image and mount the MITuna repo folder
```

File: README.md [hunk 2]
```diff
@@ -117,12 +117,21 @@ In order for a PR to be accepted the following pylint command needs to result in
...
+docker run -it -v "/path/on/host:/path/on/container" Tuna 
+
+//change directory to the mounted folder inside the docker containter
+cd ./path/on/container/MITuna 
+
+//double check that the environment is sourced. If not sourced, please source the environment as specified in the installation step
+printenv | grep TUNA
```
Comment (alexandraBara) [inline README.md]: remove, the env is set up correctly in the docker

File: vars/utils.groovy [hunk 3]
```diff
@@ -438,6 +438,29 @@ def perfEval_gfx908() {
...
+        env.TUNA_DB_PASSWORD="${db_password}"
+        env.gateway_ip = "${gateway_ip}"
+        env.gateway_port = "${gateway_port}"
+        env.gateway_user = "${gateway_user}"
+        env.TUNA_DOCKER_NAME="ci-tuna:${branch_id}_coverage"
+        env.PYTHONPATH=env.WORKSPACE
+        env.PATH="${env.WORKSPACE}/tuna:${env.PATH}"
+        addMachine(arch, num_cu, machine_ip, machine_local_ip, username, pwd, port)
+        // download the latest perf db
+        //runsql("DELETE FROM config_tags; DELETE FROM job; DELETE FROM config;")
```
Comment (alexandraBara) [inline vars/utils.groovy]: remove

File: Jenkinsfile [hunk 4]
```diff
@@ -82,7 +82,20 @@ pipeline {
...
+        }
+        stage("approx coverage"){
+        //coverage report
+        agent{  label utils.rocmnode("tunatest") }
+        steps {
+            script{
+                try {
+                    utils.coverageReport()
+                    } catch (Exception err) {
+                        currentBuild.result = 'SUCCESS'
```
Comment (alexandraBara) [inline Jenkinsfile]: this looks like we are passing it in case of exceptions. What errors get caught here?

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 4 above)
Comment (reidkwja) [reply Jenkinsfile]: For now, we are bypassing any errors that are produced by the coverage report\pytests - considering that's it's only a report showing a percentage and pytests are ran separately. In the future, we will configure it to fail if the coverage percentage drops below a certain level.

**Review by alexandraBara — APPROVED** [review]

File: Jenkinsfile [hunk 5]
```diff
@@ -82,7 +82,20 @@ pipeline {
             utils.pytestSuite3()
             }
             }
-        }    
+        }
+        stage("approx coverage"){
```
Comment (JehandadKhan) [inline Jenkinsfile]: Why approx ?

**Review by JehandadKhan — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 4 above)
Comment (alexandraBara) [reply Jenkinsfile]: @reidkwja  can you try and drop the DB and recreate it in this step? I think this should fix the DB errors that code cov throws and this catch should not be necessary.

**Review by cderb — APPROVED** [review]

File: Jenkinsfile [hunk 6]
```diff
@@ -59,6 +59,19 @@ pipeline {
             }
             }
         }
+        stage("approx coverage"){
+        //coverage report
+        agent{  label utils.rocmnode("tunatest") }
+        steps {
+            script{
+                try {
```
Comment (alexandraBara) [inline Jenkinsfile]: @reidkwja  please remove this try-catch and let it fail. Work with @msaudulhassan  to fix the failing SolverAnalytics test.

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 6 above)
Comment (msaudulhassan) [reply Jenkinsfile]: Resolved

File: Jenkinsfile [hunk 7]
```diff
@@ -42,6 +42,21 @@ pipeline {
...
+                    utils.coverageReport()
+                    } catch (Exception err) {
+                        currentBuild.result = 'SUCCESS'
+                    }
+            }
+            }
+        }
+        */
```
Comment (alexandraBara) [inline Jenkinsfile]: remove

File: tests/test_importbenchmark.py [hunk 8]
```diff
@@ -52,6 +52,7 @@ def test_import_benchmark():
   for key, value in models.items():
     args.add_model = key.value
     args.version = value
+    args.md_version = 1
```
Comment (alexandraBara) [inline tests/test_importbenchmark.py]: remove, let pylint PR take care of this

File: vars/utils.groovy [hunk 9]
```diff
@@ -454,23 +460,24 @@ def pytestSuite1() {
         addMachine(arch, num_cu, machine_ip, machine_local_ip, username, pwd, port)
         // download the latest perf db
         //runsql("DELETE FROM config_tags; DELETE FROM job; DELETE FROM config;")
-        sshagent (credentials: ['bastion-ssh-key']) {                 
-           sh "pytest tests/test_abort_file.py -s"
-           sh "pytest tests/test_analyze_parse_db.py -s"
-
-           sh "pytest tests/test_connection.py -s"
+        sshagent (credentials: ['bastion-ssh-key']) {   
+           sh "coverage erase"
```
Comment (alexandraBara) [inline vars/utils.groovy:462]: is this needed?

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: vars/utils.groovy (hunk 9 above)
Comment (reidkwja) [reply vars/utils.groovy:462]: Yes

File: .coveragerc [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+source = ./tuna
+omit = 
+      /usr/local/lib* 
+      */__init__.py
+
+[report]
+exclude_lines =
+       def __repr __
+       raise NotImplementedError
+       if __name__ == .__main__.:
```
Comment (JehandadKhan) [inline .coveragerc]: Please add a new line

**Review by JehandadKhan — CHANGES_REQUESTED** [review]
Do we need to update the `coverage.json` every time ? Is there a way to avoid storing the `coverage.json` file ?

**Comment by alexandraBara** [comment]
Here is what i think needs to also be excluded: 

- corrupt_configs.py
- data_migration_updated_layouts.py
- export_configs.py
- flask.py
- flask_example.py
- grafana_dict.py
- analyze_fdb.py
- prune_db.py
- query_db.py
- rm_dup_cfgs.py 
- rm_dup_jobs.py 
- sqlite_2txt.py 
- dupe_resolve.py
- 
Ideally we would move all these(except the flask ones) into a new folder in tuna, call it...'custom_scripts'?? and then ignore the entire folder. Don't move the flask scripts, those would need to be ignored manually

@JehandadKhan  @cderb any suggestions?

**Comment by reidkwja** [comment]
> Do we need to update the `coverage.json` every time ? Is there a way to avoid storing the `coverage.json` file ?

For manual runs locally, yes...it can be removed from the repo

**Comment by cderb** [comment]
> Ideally we would move all these(except the flask ones) into a new folder in tuna, call it...'custom_scripts'?? and then ignore the entire folder. Don't move the flask scripts, those would need to be ignored manually
> 
> @JehandadKhan @cderb any suggestions?

I'd like to go through the library again sometime soon and sort these into some tighter groupings. I'll put up a task on LWPTUNA.

**Comment by alexandraBara** [comment]
> > Ideally we would move all these(except the flask ones) into a new folder in tuna, call it...'custom_scripts'?? and then ignore the entire folder. Don't move the flask scripts, those would need to be ignored manually
> > @JehandadKhan @cderb any suggestions?
> 
> I'd like to go through the library again sometime soon and sort these into some tighter groupings. I'll put up a task on LWPTUNA.

@cderb  so for now, we just ignore them file by file from codecov?

**Comment by reidkwja** [comment]
@cderb @alexandraBara The following files have been omitted: 
```
      tuna/corrupt_configs.py
      tuna/data_migration_updated_layouts.py
      tuna/export_configs.py
      tuna/flask.py
      tuna/flask_example.py
      tuna/miopen/analyze_fdb.py
      tuna/prune_db.py
      tuna/query_db.py
      tuna/rm_dup_cfgs.py
      tuna/rm_dup_jobs.py
      tuna/sqlite_2txt.py
      tuna/utils/dupe_resolve.py
      tuna/build_driver_cmd.py
      tuna/solver_res.py   
```

**Comment by alexandraBara** [comment]
Please add a way to compare current % coverage vs  previous

**Comment by reidkwja** [comment]
> Please add a way to compare current % coverage vs previous

Done!

**Comment by alexandraBara** [comment]
> > Please add a way to compare current % coverage vs previous
> 
> Done!

So with this new script to check %coverage, we would have to update the '50' to current coverage % when it increases?

File: tests/covscripts/coverage_branches.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+coverage_data = json.load(coverage_file)
+percent_covered = coverage_data['totals']['percent_covered']
+percent_covered = '{:.2f}'.format(percent_covered)
+percent_covered = float(percent_covered)
+
+with open(file_path_prev, 'r') as f:
+  prev_coverage = f.readline().strip()
+  prev_coverage_int = float(prev_coverage)
+
+print(percent_covered)
```
Comment (alexandraBara) [inline tests/covscripts/coverage_branches.py]: Add a line of text to this number so we know its the new one

File: tests/covscripts/coverage_branches.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+print(prev_coverage_int)
```
Comment (alexandraBara) [inline tests/covscripts/coverage_branches.py]: same, add a line of text to denote that this is the old coverage

File: tests/covscripts/coverage_branches.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+if percent_covered < prev_coverage_int:
+  raise ValueError('Coverage Tests dropped below an acceptable treshold')
```
Comment (alexandraBara) [inline tests/covscripts/coverage_branches.py]:
```suggestion
  raise ValueError('Code coverage decreased')
```

File: utils/coverage_files/coverage_percentage.txt [hunk 14]
```diff
@@ -1 +1 @@
-39.39
+50.51
```
Comment (alexandraBara) [inline utils/coverage_files/coverage_percentage.txt]: is this updated automatically? or still needed?

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: utils/coverage_files/coverage_percentage.txt (hunk 14 above)
Comment (reidkwja) [reply utils/coverage_files/coverage_percentage.txt]: removing this file

File: Jenkinsfile [hunk 15]
```diff
@@ -6,6 +6,9 @@ pipeline {
     environment {
         branch =  sh(script: "echo ${scm.branches[0].name} | sed 's/[^a-zA-Z0-9]/_/g' ", returnStdout: true).trim()
         branch_id = "${branch}_${BUILD_ID}"
+        branch_dev = "develop" //will change to develop
+        branch_dev_artif = "rk_coverage_auto/lastSuccessfulBuild/artifact"
```
Comment (alexandraBara) [inline Jenkinsfile]: what happens when we delete this rk_coverage_auto branch? 
I think we should use a non-developer branch for this

Comment (reidkwja) [reply Jenkinsfile]: We will change this one to develop before merging...this is for testing purposes

File: tests/covscripts/coverage_branches.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
+###############################################################################
+#
+# MIT License
+#
+# Copyright (c) 2022 Advanced Micro Devices, Inc.
```
Comment (alexandraBara) [inline tests/covscripts/coverage_branches.py]:
```suggestion
# Copyright (c) 2023 Advanced Micro Devices, Inc.
```

File: Dockerfile [hunk 17]
```diff
@@ -9,7 +9,8 @@ ARG ROCMVERSION=5.1
 ARG OSDB_BKC_VERSION
 ARG DEB_ROCM_REPO=http://repo.radeon.com/rocm/apt/.apt_$ROCMVERSION/
 # Add rocm repository
-RUN apt-get update
+RUN apt-get update && \
+      apt-get -y install sudo
```
Comment (JehandadKhan) [inline Dockerfile]: Why do you need sudo? The docker is already running in root?

**Review by alexandraBara — APPROVED** [review]

File: Jenkinsfile [hunk 18]
```diff
@@ -1,11 +1,14 @@
...
 library "jenkins-shared@$BRANCH_NAME"
 
 pipeline {
     agent any
     environment {
         branch =  sh(script: "echo ${scm.branches[0].name} | sed 's/[^a-zA-Z0-9]/_/g' ", returnStdout: true).trim()
         branch_id = "${branch}_${BUILD_ID}"
+        branch_dev = "develop"
+        branch_dev_artif = "develop/lastSuccessfulBuild/artifact"
+        jenkin_url = "http://localhost:8080/job/Test_mb/job"
```
Comment (JehandadKhan) [inline Jenkinsfile]: Please dont hard code the URL, instead consume it from an env var

File: tests/covscripts/coverage_branches.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+###############################################################################
+import json
+import os
+
+root_dir = os.path.dirname(
+    os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
Comment (JehandadKhan) [inline tests/covscripts/coverage_branches.py]: Why are there three nested `dirname` invocations ?

File: tests/covscripts/coverage_develop.py [hunk 20]
```diff
@@ -2,7 +2,7 @@
 #
 # MIT License
 #
-# Copyright (c) 2022 Advanced Micro Devices, Inc.
+# Copyright (c) 2023 Advanced Micro Devices, Inc.
```
Comment (JehandadKhan) [inline tests/covscripts/coverage_develop.py]: Not needed, the year always marks the year a file was created

File: vars/utils.groovy [hunk 21]
```diff
@@ -525,9 +528,45 @@ def pytestSuite3() {
...
+           sh "coverage report -m"
+           sh "python3 -m coverage json"
+
+        }
+        try {
+           sh "wget http://localhost:${jenkin_port}/job/${job_name}/job/${branch_master}/${latest_succes_build}/artifact/${cov_file}"
+           sh "python3 tests/covscripts/coverage.py branch"
+        } catch (Exception err) {
+           currentBuild.result = 'SUCCESS'
+        }
```
Comment (JehandadKhan) [inline vars/utils.groovy]:
```suggestion
        try {
           sh "wget http://localhost:${jenkin_port}/job/${job_name}/job/${branch_master}/${latest_succes_build}/artifact/${cov_file}"
        } catch (Exception err) {
           currentBuild.result = 'SUCCESS'
        }
         sh "python3 tests/covscripts/coverage.py branch"
```

File: tests/covscripts/coverage.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+###############################################################################
+
+import json
+import os
+import argparse
+
+root_dir = os.path.dirname(
+    os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+file_path_curcov_json = os.path.join(root_dir, "./coverage.json")
+file_path_prevcov_txt = os.path.join(root_dir, "./develop_percent_coverage.txt")
```
Comment (JehandadKhan) [inline tests/covscripts/coverage.py]: Please pass in these names are argument

File: tests/covscripts/coverage.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    'decide whether to extract coverage for develop or a peripheral branch')
+parser.add_argument('function',
+                    choices=['develop', 'branch'],
+                    help='choose develop or branch')
+args = parser.parse_args()
+
+if args.function == 'develop':
+  develop()
+elif args.function == 'branch':
+  branch()
```
Comment (JehandadKhan) [inline tests/covscripts/coverage.py]: Its preferable to check if the file is being imported or run as script.

File: tests/covscripts/coverage.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+  else:
+    raise ValueError('Failed to save coverage artifacts')
+
+
+def branch():
+  """run and compare coverage for the non-develop, peripheral branches"""
+
+  curcov_percentage_ftdt = curcov()
+
+  with open('develop_percent_coverage.txt', 'r') as f:
```
Comment (JehandadKhan) [inline tests/covscripts/coverage.py]: Please avoid hardcoding filenames

File: vars/utils.groovy [hunk 25]
```diff
@@ -525,9 +528,46 @@ def pytestSuite3() {
...
+           sh "wget ${jenkin_url}/${branch_dev_artif}/develop_percent_coverage.txt"
+           sh "python3 tests/covscripts/coverage_branches.py"
+        } catch (Exception err) {
+           currentBuild.result = 'SUCCESS'
+        }
+    }
+}
+
+def coverageExport() {
+    def tuna_docker = docker.build("ci-tuna:${branch_id}_pytest3", " --build-arg FIN_TOKEN=${FIN_TOKEN} --build-arg BACKEND=HIP .")
```
Comment (alexandraBara) [inline vars/utils.groovy]: we should not be needing Fin token anymore

File: vars/utils.groovy (hunk 21 above)
Comment (JehandadKhan) [reply vars/utils.groovy]: If the coverage script is part of the catch and the result is ignored, then a loss of coverage will never result in failure since it would be caught and converted to success

**Review by JehandadKhan — CHANGES_REQUESTED** [review]

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: vars/utils.groovy (hunk 25 above)
Comment (reidkwja) [reply vars/utils.groovy]: Removing fin token creates build error in docker

File: tests/covscripts/coverage.py [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+  print(
+      f"Current Testing Coverage for develop branch is: {curcov_percentage_ftdt}%"
+  )
+
+  with open(COVERAGE_FILE_NAME_TXT, 'w') as f:
+    f.write(str(curcov_percentage_ftdt))
+
+  check_txt_export = check_export(COVERAGE_FILE_NAME_TXT)
+
+  if check_txt_export is True:
```
Comment (JehandadKhan) [inline tests/covscripts/coverage.py]:
```suggestion
  if check_txt_export:
```

File: tests/covscripts/coverage.py [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+
+def curcov():
+  """extract current coverage from a json file"""
+  curcov_json = open(file_path_curcov_json)
+  curcov_data = json.load(curcov_json)
+  curcov_percentage = curcov_data['totals']['percent_covered']
+  curcov_percentage_formated = '{:.2f}'.format(curcov_percentage)
+  curcov_percentage_ftdt = float(curcov_percentage_formated)
+  return curcov_percentage_ftdt
+
```
Comment (JehandadKhan) [inline tests/covscripts/coverage.py:45]:
```suggestion
def main():
  master_file = sys.argv[1]
  branch_file = sys.argv[2]
  master_pct = curcov(master_file)
  branch_pct = curcov(branch_file)
  if(branch_pct > master_pct)
    sys.exit(0)
  else
    sys.exit(-1)
```

File: tests/covscripts/coverage.py [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+import json
+import os
+import argparse
+
+COVERAGE_FILE_NAME_TXT = 'develop_percent_coverage.txt'
+COVERAGE_FILE_NAME_JSON = 'coverage.json'
+
+root_dir = os.path.dirname(__file__)
+file_path_curcov_json = os.path.join(root_dir, COVERAGE_FILE_NAME_JSON)
+file_path_prevcov_txt = os.path.join(root_dir, COVERAGE_FILE_NAME_TXT)
```
Comment (JehandadKhan) [inline tests/covscripts/coverage.py]:
```suggestion
```

File: tests/covscripts/coverage.py [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+
+def check_export(filename):
+  """check if the file was successfully exported"""
+  status = os.path.exists(filename)
+  return status
+
+
+def curcov():
+  """extract current coverage from a json file"""
+  curcov_json = open(file_path_curcov_json)
```
Comment (JehandadKhan) [inline tests/covscripts/coverage.py]:
```suggestion
def curcov(filename):
  """extract current coverage from a json file"""
  check_export(filename)
  curcov_json = open(file_path_curcov_json)
```

File: vars/utils.groovy [hunk 30]
```diff
@@ -525,9 +528,45 @@ def pytestSuite3() {
...
+           sh "coverage report -m"
+           sh "python3 -m coverage json"
+
+        }
+        try {
+           sh "wget http://localhost:${jenkin_port}/job/${job_name}/job/${branch_master}/${latest_succes_build}/artifact/${cov_file}"
+        } catch (Exception err) {
+           currentBuild.result = 'SUCCESS'
+        }
+         sh "python3 tests/covscripts/coverage.py branch"
```
Comment (JehandadKhan) [inline vars/utils.groovy]:
```suggestion
      if(git.scm.branch[0] == "${TUNA_COVERAGE_MASTER_BRANCH}")
      {
      archiveArtifacts artifacts: "coverage.json", allwEmptyArchive: true, fingerprint: true
      }
      else
      {
       try {
           sh "wget -o cov_file_from_develop.json ${env.TUNA_COVERAGE_URL}/lastSuccessfulBuild/artifact/${cov_file}"
        } catch (Exception err) {
           currentBuild.result = 'SUCCESS'
        }
         sh "python3 tests/covscripts/coverage.py conv_file_from_develop.json coverage.json"
      }
       
```

File: vars/utils.groovy [hunk 31]
```diff
@@ -525,9 +528,45 @@ def pytestSuite3() {
...
+        //runsql("DELETE FROM config_tags; DELETE FROM job; DELETE FROM config;")
+        sshagent (credentials: ['bastion-ssh-key']) {                 
+           // test fin builder and test fin builder conv in sequence
+           sh "python3 -m coverage run -a -m pytest tests/test_fin_evaluator.py -s"
+           sh "python3 -m coverage run -a -m pytest tests/test_update_golden.py -s"
+           sh "coverage report -m"
+           sh "python3 -m coverage json"
+           sh "python3 tests/covscripts/coverage.py develop"
         }
+        archiveArtifacts artifacts: "develop_percent_coverage.txt", allwEmptyArchive: true, fingerprint: true
```
Comment (JehandadKhan) [inline vars/utils.groovy]:
```suggestion
```

File: vars/utils.groovy (hunk 30 above)
Comment (JehandadKhan) [reply vars/utils.groovy]: Also, please check  the spelling of `allwEmptyArchive`

File: Jenkinsfile [hunk 32]
```diff
@@ -76,13 +81,18 @@ pipeline {
...
-        agent{  label utils.rocmnode("tunatest") }
+        agent{ label utils.rocmnode("tunatest") }
         steps{
             script{
-            utils.pytestSuite3()
+                if (branch == branch_master) { 
+                utils.coverageExport()
+                } else {
+                utils.pytestSuite3()
+                }
```
Comment (JehandadKhan) [inline Jenkinsfile]: This conditional should be moved to the respective function

**Review by JehandadKhan — DISMISSED** [review]

File: Jenkinsfile [hunk 33]
```diff
@@ -42,6 +44,7 @@ pipeline {
            }
            }
         }
+        **/
```
Comment (reidkwja) [inline Jenkinsfile]: Adding this back to get that step to pass with recent restructuring of files.

File: vars/utils.groovy [hunk 34]
```diff
@@ -575,6 +575,7 @@ def runLint() {
             sh "mypy tuna/worker_interface.py --ignore-missing-imports --follow-imports=skip"
             sh "yamllint tuna/miopen/yaml_files/*.yaml"
             sh "yamllint tuna/example/*.yaml"
+            sh "mypy tuna/miopen/driver/base.py --ignore-missing-imports --follow-imports=skip"
```
Comment (alexandraBara) [inline vars/utils.groovy]: why did you add this here?

File: tests/covscripts/coverage.py (hunk 27 above)
Comment (reidkwja) [reply tests/covscripts/coverage.py:45]: I re-implemented similar logic in the parse args - allowing it to take file names as an argument. Would that be okay?

File: vars/utils.groovy [hunk 35]
```diff
@@ -520,14 +523,25 @@ def pytestSuite3() {
...
+        }
+        if (current_run == main_branch) {
+            sh "python3 tests/covscripts/coverage.py ${main_branch}"
+            archiveArtifacts artifacts: "${env.COVERAGE_ARTIFACT_FILE_NAME}", allowEmptyArchive: true, fingerprint: true
+        } else {
+        try {
+            sh "wget ${env.TUNA_COVERAGE_URL}/${main_branch}/lastSuccessfulBuild/artifact/${env.COVERAGE_ARTIFACT_FILE_NAME}"
+            sh "python3 tests/covscripts/coverage.py ${current_run}"
+        } catch (Exception err) {
+            currentBuild.result = 'SUCCESS'
```
Comment (alexandraBara) [inline vars/utils.groovy:538]: @reidkwja when does this fail? and does it fail if the coverage decreased?

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: vars/utils.groovy (hunk 35 above)
Comment (reidkwja) [reply vars/utils.groovy:538]: Changed back...was there for testing the branch

File: README.md [hunk 36]
```diff
@@ -116,5 +116,6 @@ Static code analysis
 In order for a PR to be accepted the following pylint command needs to result in 10/10 analysis:
 ```
 cd MITuna/tuna
+pylint -f parseable -d duplicate-code --max-args=8 --indent-string '  ' *.py
```
Comment (msaudulhassan) [inline README.md]: Wouldn't the next line do this?

**Review by cderb — APPROVED** [review]

File: Jenkinsfile [hunk 37]
```diff
@@ -75,14 +77,14 @@ pipeline {
...
-        agent{  label utils.rocmnode("tunatest") }
-        steps{
-            script{
-            utils.pytestSuite3()
-            }
+        stage("pytest3 and Tests Coverage"){
+            agent { label utils.rocmnode("tunatest") }
+            steps {
+                script {
+                    utils.pytestSuit3AndCoverage(branch, branch_master)
```
Comment (msaudulhassan) [inline Jenkinsfile]: I might be missing something, but can't we put the generation of coverage report into a separate stage?

File: tests/covscripts/coverage.py [hunk 38]
```diff
@@ -0,0 +1,118 @@
...
+
+  print(
+      f"Current Testing Coverage for current branch is: {curcov_percentage_ftdt}%"
+  )
+  print(
+      f"Current Testing Coverage for develop branch is: {prevcov_percentage_ftdt}%"
+  )
+  if curcov_percentage_ftdt < prevcov_percentage_ftdt:
+    raise ValueError('Code Coverage Decreased')
+
```
Comment (msaudulhassan) [inline tests/covscripts/coverage.py:88]: Since the coverage percentage is already in the json file, why do we need the txt file needed?

File: tests/covscripts/coverage.py [hunk 39]
```diff
@@ -0,0 +1,118 @@
...
+  """extract current coverage from a json file"""
+  with open(file_path_curcov_json) as curcov_json:
+    curcov_data = json.load(curcov_json)
+  curcov_percentage = curcov_data['totals']['percent_covered']
+  curcov_percentage_ftdt = round(curcov_percentage, 2)
+  return curcov_percentage_ftdt
+
+
+def develop(root_dir, coverage_file_name_txt, coverage_file_name_json):
+  """run and compare coverage for develop branch"""
```
Comment (msaudulhassan) [inline tests/covscripts/coverage.py]: Is this function just creating the .txt file and not doing any comparing?

Comment (msaudulhassan) [reply tests/covscripts/coverage.py]: And if it is comparing branches, which branches are being compared here? develop to what?

**Review by msaudulhassan — CHANGES_REQUESTED** [review]
Minor comments

File: README.md (hunk 36 above)
Comment (reidkwja) [reply README.md]: I think this accidentally flagged as an edit...removed manual processes

File: tests/covscripts/coverage.py (hunk 39 above)
Comment (reidkwja) [reply tests/covscripts/coverage.py]: This is tailored for exporting the txt file out when develop is ran...no comps

File: tests/covscripts/coverage.py (hunk 38 above)
Comment (reidkwja) [reply tests/covscripts/coverage.py:88]: Yeah, the thing about JSON file is that its larger and contains more data, and vars are parsed out regardless...txt file is cleaner\smaller

File: Jenkinsfile (hunk 37 above)
Comment (reidkwja) [reply Jenkinsfile]: thought about doing that...opted to keep it with pytest3suite

File: tests/covscripts/coverage.py (hunk 39 above)
Comment (msaudulhassan) [reply tests/covscripts/coverage.py]: Gotcha, it'd help if the function name/docstring makes this clear. Also, the if-else checks to see if coverage_file_name_txt got created successfully -- are they necessary? I don't see why the file would not get created.

**Review by alexandraBara — APPROVED** [review]

File: README.md [hunk 40]
```diff
@@ -116,5 +116,6 @@ Static code analysis
...
 pylint -f parseable -d duplicate-code --max-args=8 --indent-string '  ' miopen/*.py example/*.py *.py
```
Comment (msaudulhassan) [inline README.md]: Please take out L119 since it's doing the same thing as L120 is already doing

File: tests/covscripts/coverage.py (hunk 27 above)
Comment (msaudulhassan) [reply tests/covscripts/coverage.py:45]: I think it's better to return an exit code like JD (in https://github.com/ROCmSoftwarePlatform/MITuna/pull/836/files/b6eb4ca14e124fde9d3255a8b4d28a48e97b8467#r1137621287) is doing instead of raising an exception if branch_pct < master_pct.

File: Jenkinsfile (hunk 37 above)
Comment (msaudulhassan) [reply Jenkinsfile]: I'm not sure how putting them in a single stage helps? Since they are two separate tests, wouldn't it be better to carry them out in separate stages? If the coverage test fails, you can just rerun the coverage stage, instead of having to rerun both PytestSuite3 and Coverage.

(Also, there seems to be a slight misspelling in pytestSuit3)

**Review by msaudulhassan — CHANGES_REQUESTED** [review]

File: README.md (hunk 36 above)
Comment (msaudulhassan) [reply README.md]: See my later comment on this

File: Jenkinsfile (hunk 37 above)
Comment (reidkwja) [reply Jenkinsfile]: Spelling Fixed. The rest is feasible for a separate PR for the current timing.

File: README.md (hunk 40 above)
Comment (reidkwja) [reply README.md]: Fixed

**Comment by reidkwja** [comment]
Additional suggestions by @msaudulhassan can be addressed in future changes.

File: README.md (hunk 36 above)
Comment (reidkwja) [reply README.md]: done!

**Review by msaudulhassan — APPROVED** [review]
